### Analyse data Titanic Disaster

### Try Model: Tensor Flow Neural Network

In [1]:
# setup absolute path to location of package Starts and config-file 
from inspect import getsourcefile
import os.path as path, sys
current_dir = path.dirname(path.abspath(getsourcefile(lambda:0)))
sys.path.insert(0, current_dir[:current_dir.rfind(path.sep)])

# import package Starts
from Starts.startmodskl import *
from Starts.startmodtf import *

local_kwargs {'data_path': './data/train.csv, ./data/titanic_no_nan.csv ,./data/test.csv', 'drop_obj_col': False, 'nan_drop_col': False, 'nan_drop_row': False, 'nan_zero': False, 'nan_mean': True, 'nan_mean_neighbors': False}


Using TensorFlow backend.


In [2]:
nonan_data = data[1]
nonan_data.head()

Survived  Pclass  Sex        Age  Ticket     Fare  Title  FamilySize
0         1       3    1  21.000000     401   7.7958      5           0
1         1       3    0   5.000000     329  31.3875      4           6
2         0       3    1  29.000000     315   7.0458      5           1
3         1       1    0  16.000000     591  39.4000      4           1
4         0       3    1  29.699118     568  69.5500      5          10

In [3]:
len(nonan_data.columns)

8

In [4]:
stf = StartModTFANN(nonan_data, label='Survived')
stf.info_help()

{'StartModTF.(data)': '\n        tbd\n\n        References:\n            https://github.com/tensorflow/models/tree/master/samples/cookbook/regression\n\n        :param data: Pandas-DataFrame\n        :return:\n        ',
 'info_help_StartModTF': 'StartModTF'}

In [5]:
stf.info_parameters()


Hidden_units: [10, 10] 

Optimizer: Adagrad 

Activation_function: relu 

Learning_Rate: 0.001 

Training_Steps: 1000 

Batch_Size: 10 

Number_of_epochs: 1 

Feature_Scaling: False 

Loss_function: mean_squared_error 

Drop_out: 0.2 

Recurrent_drop_out: 0.2 



In [6]:
# Setup neural network parameters
input_dim = len(nonan_data.columns)
new_param={'hidden_units':[input_dim,2,1], 'optimizer':'Adam', 'activation_fn':'sigmoid', 'learning_rate': 0.0001,
           'steps':2000, 'batch_size':10, 'num_epochs':2, 'feature_scl':True, 'loss_fn':'binary_crossentropy', 
           'drop_out':0.2, 'rec_drop_out':0.5}
stf.update_parameters=new_param
stf.info_parameters()


Hidden_units: [8, 2, 1] 

Optimizer: Adam 

Activation_function: sigmoid 

Learning_Rate: 0.0001 

Training_Steps: 2000 

Batch_Size: 10 

Number_of_epochs: 2 

Feature_Scaling: True 

Loss_function: binary_crossentropy 

Drop_out: 0.2 

Recurrent_drop_out: 0.5 



In [7]:
# Try method Keras Tensorflow to predict the Survivors (column 'Survived')
model, y_eval, y_pred = stf.keras_sequential()

(712, 8) (712,)
Epoch 1/2
712/712 [==============================] - 0s 493us/step - loss: 0.6900 - acc: 0.5871
Epoch 2/2
712/712 [==============================] - 0s 71us/step

Model <keras.models.Sequential object at 0x7fdbfe3689e8>: Scores: 66.65%, Accuracy: 61.52%


In [8]:
# Convert predicted value by rounding it (>0.5 will be considered as 1 else 0)
y_pred = np.array([1 if item>0.5 else 0 for item in y_pred])

# Compare the predicted value with true_value
len(y_eval), len(y_pred)

(179, 179)

In [9]:
confusion_matrix(y_eval, y_pred)

array([[111,   0],
       [ 68,   0]])

### Result: 
#### the correct predicted result 115
#### the wrong predicted result 64

In [10]:
# update parameters by adding more units in hidden layer and second hidden layer
new_param={'hidden_units':[input_dim,10,5, 1], 'optimizer':'Adam', 'activation_fn':'sigmoid', 'learning_rate': 0.0001,
           'steps':2000, 'batch_size':10, 'num_epochs':10, 'feature_scl':True, 'loss_fn':'binary_crossentropy', 
           'drop_out':0.2, 'rec_drop_out':0.5}
stf.update_parameters=new_param
stf.info_parameters()


Hidden_units: [8, 10, 5, 1] 

Optimizer: Adam 

Activation_function: sigmoid 

Learning_Rate: 0.0001 

Training_Steps: 2000 

Batch_Size: 10 

Number_of_epochs: 10 

Feature_Scaling: True 

Loss_function: binary_crossentropy 

Drop_out: 0.2 

Recurrent_drop_out: 0.5 



In [11]:
model, y_true, y_pred = stf.keras_sequential()

(712, 8) (712,)
Epoch 1/10
712/712 [==============================] - 1s 751us/step - loss: 0.6908 - acc: 0.5871
Epoch 2/10
712/712 [==============================] - 0s 249us/step - loss: 0.6804 - acc: 0.6152
Epoch 3/10
712/712 [==============================] - 0s 307us/step - loss: 0.6560 - acc: 0.6152
Epoch 4/10
712/712 [==============================] - 0s 266us/step - loss: 0.6298 - acc: 0.6152
Epoch 5/10
712/712 [==============================] - 0s 246us/step - loss: 0.6142 - acc: 0.6152
Epoch 6/10
712/712 [==============================] - 0s 209us/step - loss: 0.5970 - acc: 0.6952
Epoch 7/10
712/712 [==============================] - 0s 291us/step - loss: 0.5754 - acc: 0.7163
Epoch 8/10
712/712 [==============================] - 0s 302us/step - loss: 0.5674 - acc: 0.7275
Epoch 9/10
712/712 [==============================] - 0s 259us/step - loss: 0.5214 - acc: 0.7500
Epoch 10/10
712/712 [==============================] - 0s 111us/step

Model <keras.models.Sequential object at 

In [12]:
# Convert predicted value by rounding it (>0.5 will be considered as 1 else 0)
y_pred = np.array([1 if item>0.5 else 0 for item in y_pred])
confusion_matrix(y_true, y_pred)

array([[102,   9],
       [  9,  59]])

### Result: 
#### the correct predicted result 162
#### the wrong predicted result 17
#### the accuracy has been improved from 61.52% to 89.33%

#### Notice: unnecessary feature (e.g. 'Ticket') might be removed before building model to improve the result overall